RecoLocalCalo/HcalRecProducers/src/MahiGPU.cu

```diff
@@ -0,0 +1,1403 @@
+      static_assert(NPULSES == NSAMPLES);
+
+      // indices
+      auto const gch = threadIdx.x + blockIdx.x * blockDim.x;
+      auto const nchannelsf015 = nchannelsf01HE + nchannelsf5HB;
+      if (gch >= nchannelsTotal)
+        return;
+
+      // if chi2 is set to -9999 do not run minimization
+      if (outputChi2[gch] == -9999.f)
+        return;
+
+      // configure shared mem
+      extern __shared__ char shrmem[];
+      float* shrMatrixLFnnlsStorage =
+          reinterpret_cast<float*>(shrmem) + calo::multifit::MapSymM<float, NPULSES>::total * threadIdx.x;
+      float* shrAtAStorage = reinterpret_cast<float*>(shrmem) +
+                             calo::multifit::MapSymM<float, NPULSES>::total * (threadIdx.x + blockDim.x);
+
+      // conditions for pedestal widths
+      auto const id = gch < nchannelsf01HE
+                          ? idsf01HE[gch]
+                          : (gch < nchannelsf015 ? idsf5HB[gch - nchannelsf01HE] : idsf3HB[gch - nchannelsf015]);
+      //auto const id = gch >= nchannelsf01HE
+      //    ? idsf5HB[gch - nchannelsf01HE]
+      //    : idsf01HE[gch];
+      auto const did = DetId{id};
+      auto const hashedId =
+          did.subdetId() == HcalBarrel
+              ? hcal::reconstruction::did2linearIndexHB(id, maxDepthHB, firstHBRing, lastHBRing, nEtaHB)
+              : hcal::reconstruction::did2linearIndexHE(id, maxDepthHE, maxPhiHE, firstHERing, lastHERing, nEtaHE) +
+                    offsetForHashes;
+
+      auto const* pedestalWidthsForChannel = useEffectivePedestals && (gch < nchannelsf01HE || gch >= nchannelsf015)
+                                                 ? effectivePedestalWidths + hashedId * 4
+                                                 : pedestalWidths + hashedId * 4;
+      auto const averagePedestalWidth2 = 0.25 * (pedestalWidthsForChannel[0] * pedestalWidthsForChannel[0] +
+                                                 pedestalWidthsForChannel[1] * pedestalWidthsForChannel[1] +
+                                                 pedestalWidthsForChannel[2] * pedestalWidthsForChannel[2] +
+                                                 pedestalWidthsForChannel[3] * pedestalWidthsForChannel[3]);
+      auto const* gains = gainValues + hashedId * 4;
+      // FIXME on cpu ts 0 capid was used - does it make any difference
+      auto const gain = gains[0];
+      auto const respCorrection = respCorrectionValues[hashedId];
+
+#ifdef HCAL_MAHI_GPUDEBUG
+#ifdef HCAL_MAHI_GPUDEBUG_FILTERDETID
+      if (id != DETID_TO_DEBUG)
+        return;
+#endif
+#endif
+
+      /*
+      // TODO: provide this properly
+      int const soi = soiSamples[gch];
+      */
+      constexpr float deltaChi2Threashold = 1e-3;
+
+      calo::multifit::ColumnVector<NPULSES, int> pulseOffsets;
+#pragma unroll
+      for (int i = 0; i < NPULSES; ++i)
+        pulseOffsets(i) = i;
+      //        pulseOffsets(i) = pulseOffsetValues[i] - pulseOffsetValues[0];
```

Comment:
Please remove commented out lines, if not needed; or alternatively add another comment line which explains why they must be kept here